### PR TITLE
Add OpApplyPrimArg and OpApplyPrimReg

### PR DESCRIPTION
**Pull request by tymm**
In Rosette compiling the expression `(fx+ 1 (fx+ 2 3))` results in the following `Code` object:
```
litvec:
  0:   {RequestExpr}
codevec:
  0:   alloc 2
  1:   lit 2,arg[0]
  2:   lit 3,arg[1]
  3:   fx+ 2,arg[1]
  5:   lit 1,arg[0]
  6:   fx+ 2,rslt
  8:   rtn/nxt
```
`fx+ 2,arg[1]` corresponds to `OpApplyPrimArg(unwind = false, next = false, nargs = 2, primNum = 226, arg = 1)` in Roscala and `fx+ 2,rslt` corresponds to `OpApplyPrimReg(unwind = false, next = false, nargs = 2, primNum = 226, reg = 0)`.

Both opcodes run a primitive on the arguments given in `state.ctxt.argvec`. In this example the `fx+` primitive is run (has the number `226`). `OpApplyPrimArg` copies the primitive result to a position in `state.ctxt.argvec` (here to the 1st position) while `OpApplyPrimReg` copies the result to a register in `state.ctxt` (here the `rslt` register). 

The above example can be observed in gdb by setting the following breakpoint: `b src/Vm.cc:795 if code->codevec->instr(0)->word == 1026`

### CODE REVIEW

File: roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala [hunk 1]
```diff
@@ -41,6 +42,33 @@ object VirtualMachine {
...
+      // TODO: Port handleException for OTsysval
+      case Left(_) => modify(_.copy(doNextThreadFlag = true))
+    }
+
+  def runPrim(unwind: Boolean, optPrim: Option[Prim]): State[VMState, Result] =
+    State { state =>
+      optPrim match {
+        case Some(prim) =>
+          if (unwind) {
+            val (res, st) = unwindAndApplyPrim(prim, state)
```
Comment (DoSOfRedRiver) [inline roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala]: `unwindAndApplyPrim(prim, state).swap`

File: roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala [hunk 2]
```diff
@@ -829,4 +853,11 @@ object VirtualMachine {
 
   def execute(op: OpUnknown, state: VMState): VMState =
     state.set(_ >> 'exitFlag)(true).set(_ >> 'exitCode)(1)
+
+  def inspect[A](f: VMState => A): State[VMState, A] =
+    State.inspect[VMState, A](f)
```
Comment (DoSOfRedRiver) [inline roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala]: Is there any reason to not use a curried form? 
`val inspect = State.inspect[String,Int] _`

File: roscala/src/main/scala/coop/rchain/rosette/prim/Prim.scala [hunk 3]
```diff
@@ -26,6 +26,8 @@ abstract class Prim extends Ob {
     val n = ctxt.nargs
 
     if (minArgs <= n && n <= maxArgs) {
+      println(fn(ctxt))
+      println(this.name)
```
Comment (DoSOfRedRiver) [inline roscala/src/main/scala/coop/rchain/rosette/prim/Prim.scala]: It looks like some debug code got here.

Comment (tymm) [reply roscala/src/main/scala/coop/rchain/rosette/prim/Prim.scala]: Thanks.

File: roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala [hunk 4]
```diff
@@ -517,66 +548,59 @@ object VirtualMachine {
...
-        val prim  = Prim.nthPrim(op.primNum)
-        val regno = op.reg
+    st.runS(state).value
+  }
 
-        val (result, newState) =
-          if (op.unwind) { unwindAndApplyPrim(prim.get, state) } else {
-            // TODO: Fix
-            (prim.get.dispatchHelper(state.ctxt), state)
+  def execute(op: OpApplyPrimReg, state: VMState): VMState = {
```
Comment (DoSOfRedRiver) [inline roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala]: Personally, I would prefer to replace such signatures with `State[VMState,Unit]`. So, why don't we go for `State`?

Comment (tymm) [reply roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala]: Yes, sure. But we would have to rewrite all `execute`s. That's something for another PR.

**Pull request merged**